### PR TITLE
BUG: stats: Fix for gh-17444.

### LINKED ISSUE #17444: BUG: beta.ppf causes segfault
### Describe your issue.

Calling `beta.ppf` with extreme values causes an segfault.

This issues may be related to the handling of `boost::wrapexcept<std::overflow_error>` exceptions:
For the following SciPy/NumPy/Python version: `1.7.0 1.19.5 sys.version_info(major=3, minor=8, micro=10, releaselevel='final', serial=0)` on gets the following error instead:
```
terminate called after throwing an instance of 'boost::wrapexcept<std::overflow_error>'
  what():  Error in function boost::math::beta<double>(double,double): numeric overflow
Aborted (core dumped)
```

This may be related to #14704 or even #14606.

### Reproducing Code Example

```python
from scipy.stats import beta
beta.ppf(.9, 1e-320, 1234)
```

### Error message

```shell
Segmentation fault (core dumped)
```

### SciPy/NumPy/Python version information

1.9.3 1.23.4 sys.version_info(major=3, minor=10, micro=6, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
The argument `message` of the user defined error handler in the stats Boost wrapper can be NULL. The code didn't check for a null pointer, so it would trigger a segmentation fault when it used `message`.

Closes gh-17444.

### CODE REVIEW

**Review by mdhaber — APPROVED** [review]
I'm not great with C++, but this is simple enough that I think I can review it. I'll test locally, and if everything comes back green, I'll merge.

_Update: locally, the `OverflowError` and its stack trace looks good to me._

<details>

```
stats.beta.ppf(p, a, b)
Traceback (most recent call last):

  File "C:\Users\matth\AppData\Local\Temp\ipykernel_27988\1477691096.py", line 1, in <cell line: 1>
    stats.beta.ppf(p, a, b)

  File "C:\Users\matth\scipy\scipy\stats\_distn_infrastructure.py", line 2359, in ppf
    place(output, cond, self._ppf(*goodargs) * scale + loc)

  File "C:\Users\matth\scipy\scipy\stats\_continuous_distns.py", line 694, in _ppf
    return _boost._beta_ppf(q, a, b)

OverflowError: Error in function boost::math::tgamma<double>(double): Overflow Error
```

</details>

**Comment by WarrenWeckesser** [comment]
Thanks @mdhaber.  This is a minimalist fix to prevent the seg. fault.  A possible follow-up is to handle subnormal inputs in a friendlier way--which might mean just providing a nicer error message--but that will be pretty far down on my own to-do list.

**Comment by mdhaber** [comment]
It seems sufficient to me for now... and for a long time.
Unfortunately, in Main Linux_Python_38_32bit_full the error doesn't get raised.

```

        a_is_subnormal = True
        b          = 12.5
        method     = <bound method rv_continuous.isf of <scipy.stats._continuous_distns.beta_gen object at 0xeaf41070>>
        p          = 0.9
        self       = <scipy.stats.tests.test_distributions.TestBeta object at 0xe2df4a30>
_____________ TestBeta.test_beta_ppf_with_subnormal_a_b[False-ppf] _____________
[gw0] linux -- Python 3.8.0 /usr/bin/python3.8
/usr/local/lib/python3.8/dist-packages/scipy-1.10.0.dev0+unknown.unknown-py3.8-linux-i686.egg/scipy/stats/tests/test_distributions.py:3383: in test_beta_ppf_with_subnormal_a_b
    stats.beta.ppf(p, a, b)
E   Failed: DID NOT RAISE <class 'OverflowError'>
        a          = 12.5
        a_is_subnormal = False
        b          = 1e-310
        method     = <bound method rv_continuous.ppf of <scipy.stats._continuous_distns.beta_gen object at 0xeaf41070>>
        p          = 0.9
        self       = <scipy.stats.tests.test_distributions.TestBeta object at 0xe2df48e0>
_____________ TestBeta.test_beta_ppf_with_subnormal_a_b[False-isf] _____________
[gw0] linux -- Python 3.8.0 /usr/bin/python3.8
/usr/local/lib/python3.8/dist-packages/scipy-1.10.0.dev0+unknown.unknown-py3.8-linux-i686.egg/scipy/stats/tests/test_distributions.py:3383: in test_beta_ppf_with_subnormal_a_b
    stats.beta.ppf(p, a, b)
E   Failed: DID NOT RAISE <class 'OverflowError'>
        a          = 12.5
        a_is_subnormal = False
        b          = 1e-310
        method     = <bound method rv_continuous.isf of <scipy.stats._continuous_distns.beta_gen object at 0xeaf41070>>
        p          = 0.9
        self       = <scipy.stats.tests.test_distributions.TestBeta object at 0xe2df4790>
--------- generated xml file: /scipy/build/test/junit/test-results.xml ---------
```

**Comment by WarrenWeckesser** [comment]
I know what is happening there: on 32 bit platforms, the Boost policy of internally promoting `double` to `long double` is enabled, and in that case, the overflow doesn't occur.

Something seems not quite right about setting that promotion policy based on whether or not the platform is 32 bit.  Why are we determining *floating point* behavior based on this?  I'd like to get that resolved (even if that only means that I get a better understanding of why we're doing that) before making any additional tweaks here.  I'll open up a new issue about it.

**Comment by mdhaber** [comment]
OK, but let's aim to get this in before 1.10 branches.

**Comment by mckib2** [comment]
> I know what is happening there: on 32 bit platforms, the Boost policy of internally promoting `double` to `long double` is enabled, and in that case, the overflow doesn't occur.
> 
> Something seems not quite right about setting that promotion policy based on whether or not the platform is 32 bit. Why are we determining _floating point_ behavior based on this? I'd like to get that resolved (even if that only means that I get a better understanding of why we're doing that) before making any additional tweaks here. I'll open up a new issue about it.

gh-17482 is now merged

**Review by mckib2 — APPROVED** [review]

**Comment by WarrenWeckesser** [comment]
I have updated the test and all the jobs are now passing.

File: scipy/stats/tests/test_distributions.py [hunk 1]
```diff
@@ -3388,6 +3389,30 @@ def test_boost_eval_issue_14606(self):
...
+        # would result in a segmentation fault if either a or b was subnormal.
+        p = 0.9
+        # Depending on the version of Boost that we have vendored and
+        # our setting of the Boost double promotion policy, the call
+        # `stats.beta.ppf(p, a, b)` might raise an OverflowError or
+        # return a value.  We'll accept either behavior (and not care about
+        # the value), because our goal here is to verify that the call does
+        # not trigger a segmentation fault.
+        try:
+            stats.beta.ppf(p, a, b)
```
Comment (tylerjereddy) [inline scipy/stats/tests/test_distributions.py:3405]: Shouldn't this be changed to the `method` `pytest` parameter?

Apart from that, this should be an easy decision to merge--fixes segfault, two core dev approvals, CI passing, small patch with test, etc.

Comment (WarrenWeckesser) [reply scipy/stats/tests/test_distributions.py]: Yes indeed it should!  I pushed an update; tests are running again.

**Comment by tylerjereddy** [comment]
Alright, let's put this in, thanks

**Pull request merged**